### PR TITLE
ImportC: add support for importing modules with __import

### PR DESCRIPTION
**Pull request by WalterBright**
I still can't believe I made this work with 10 lines of code.

### CODE REVIEW

File: changelog/__import.dd [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+---
+__import core.stdc.stdarg; // import a D source file
+__import test2;		   // import an ImportC file
+
+int foo()
+{
+    va_list x;
+    return 1 + A;
+}
```
Comment (ibuclaw) [inline changelog/__import.dd:13]: It's not clear from the changelog that this is C source code.

File: changelog/__import.dd [hunk 2]
```diff
@@ -0,0 +1,21 @@
+Add ability to import modules to ImportC
+
+ImportC can now import modules with the __import keyword, with the C code:
+
+---
+__import core.stdc.stdarg; // import a D source file
+__import test2;		   // import an ImportC file
```
Comment (kinke) [inline changelog/__import.dd:7]: So we have `betterC` as crippled D and `ImportC` as C on steroids? ;)

**Comment by maxhaton** [comment]
What is this for?

**Comment by 12345swordy** [comment]
This PR needs a rational.

**Comment by WalterBright** [comment]
> This PR needs a rational.

It completes the circle. D can import C code. C can import D code. And if you're using ImportC, you won't need `dtoh`.

File: changelog/__import.dd (hunk 2 above)
Comment (WalterBright) [reply changelog/__import.dd:7]: yup! Use ImportC, and you won't need `dtoh` anymore!

**Review by 12345swordy — APPROVED** [review]
I have no personal objections to this PR.

**Comment by andralex** [comment]
This is up there with the most awesome hacks I've ever seen.

**Comment by 12345swordy** [comment]
Side note: does this mean you are planing to roll out with a custom made preprocessor for dmd?

File: changelog/__import.dd [hunk 3]
```diff
@@ -0,0 +1,22 @@
+Add ability to import modules to ImportC
+
+ImportC can now import modules with the __import keyword, with the C code:
+
+$(CCODE
+__import core.stdc.stdarg; // import a D source file
+__import test2;		   // import an ImportC file
```
Comment (thewilsonator) [inline changelog/__import.dd:7]:
```
Searching for tabs
./changelog/pending.dd:59:__import test2;		   // import an ImportC file
```

File: changelog/__import.dd (hunk 2 above)
Comment (maxhaton) [reply changelog/__import.dd]: > yup! Use ImportC, and you won't need `dtoh` anymore!

`dtoh` exports a templated struct to mimic a D slice, how does this replicate that?

File: changelog/__import.dd (hunk 3 above)
Comment (WalterBright) [reply changelog/__import.dd]: fixed

File: test/compilable/cimport.c [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// https://issues.dlang.org/show_bug.cgi?id=22666
+
+__import core.stdc.stdarg;
+__import imports.impcimport;
+
+int foo()
+{
+    va_list x;
```
Comment (maxhaton) [inline test/compilable/cimport.c:8]: Whatabout a non-trivial D symbol like `va_arg`?

Comment (WalterBright) [reply test/compilable/cimport.c:8]: It's just another symbol.

Comment (WalterBright) [reply test/compilable/cimport.c:8]: Er, no it's not. va_arg is a template. The way to deal with that is to import another D module that instantiates the va_arg template and presents it to ImportC as a global declaration of a function.

**Review by RazvanN7 — APPROVED** [review]

**Comment by maxhaton** [comment]
The compiler currently has special cases for importing C code into/as D code, are we now going to have special cases for importing D into C into D?

**Comment by ljmf00** [comment]
I have no objection to this. I think this can have good use of D in C, although all identifiers starting with `__` are technically reserved, according to the C standard and the `import` keyword is too generic. What I mean by that is that in the future this can potentially create conflicts, especially if some compiler wants to extend C++20 `import` to C as some sort of extension or use it internally. I suggest `__importd`, `__dimport` or `__d_import` alternatives to avoid/mitigate it.

**Comment by WalterBright** [comment]
> are we now going to have special cases for importing D into C into D?

If necessary.

> What I mean by that is that in the future this can potentially create conflicts

The potential of conflicting keywords is always going to be there. We'll just deal with it when and if specific issues come up.

**Comment by ljmf00** [comment]
> > What I mean by that is that in the future this can potentially create conflicts
> 
> The potential of conflicting keywords is always going to be there. We'll just deal with it when and if specific issues come up.

It is harder that way. If that happens, since this is publicly known, the deprecation process needs extra caution and it is always tedious for the end-user to deal with such deprecations.

**Pull request merged**